### PR TITLE
CI testing improvements

### LINKED ISSUE #2515: [CI] Use official Helm chart-testing
**What problem are you trying to solve?**
Helm has its own official chart testing action, based upon the eponymous CLI tool, called [chart-testing](https://github.com/helm/chart-testing-action) which may be good to standardize on.

**Describe the solution you'd like**

Try and adopt chart-testing in CI tests.

**Describe alternatives you've considered**

Use existing which may have gaps.

**How would users interact with this feature?**

Internal only.

### LINKED ISSUE #2708: Helm tests for aggregator
We need to start building a matrix of tests for the Helm chart. Since the aggregator is a new component and an important configuration, we could start with below as suggested.

```yaml
kubecostAggregator:
  enabled: true
  cloudCost:
    enabled: true
kubecostModel:
  federatedStorageConfigSecret: federated-store
kubecostProductConfigs:
  cloudIntegrationSecret: cloud-integration
```

_Originally posted by @thomasvn in https://github.com/kubecost/cost-analyzer-helm-chart/issues/2699#issuecomment-1786201083_

### PR DESCRIPTION
**Pull request by chipzoller**
## What does this PR change?

Closes #2708 
Closes #2515 

This PR dramatically expands and improves chart testing in this repository by setting up an extensible framework, documenting it, and adding 3 files which cover most of the common Kubecost deployment configurations. For further details on this PR, see the DEVELOPMENT.md guide contained herein.

## Does this PR rely on any other PRs?

No

## How does this PR impact users? (This is the kind of thing that goes in release notes!)

No user impact. This is internal only.

## What risks are associated with merging this PR? What is required to fully test this PR?

CI could break.

## How was this PR tested?

Extensive testing in fork.

## Have you made an update to documentation? If so, please provide the corresponding PR.

Developer docs updated in DEVELOPMENT.md

### CODE REVIEW

File: .github/workflows/chart.yaml [hunk 1]
```diff
@@ -60,19 +72,65 @@ jobs:
...
+
+      - name: Uninstall chart
+        run: helm uninstall kubecost -n kubecost --no-hooks --wait
+
+      - name: Cleanup all Pods in Kubecost Namespace
+        run: kubectl -n kubecost delete deployments,daemonsets,statefulsets,pods --all --force
+
+      # Loops over all the other values files in the ci directory and installs the chart, runs Helm tests, and uninstalls.
+      # Additional sleeps seem necessary for the main Kubecost Pod to be completely available, both frontend and backend.
+      - name: Install, test, remove chart for other values
```
Comment (thomasvn) [inline .github/workflows/chart.yaml:113]: Could you help clarify how this test, which installs all yaml files in `cost-analyzer/ci` one by one, differs from the tests run above? Are the tests run above for a default install of Kubecost?

File: .github/workflows/chart.yaml [hunk 2]
```diff
@@ -60,19 +72,65 @@ jobs:
...
+            echo "### Performing installation with values from $file ###"
+            helm install --wait --wait-for-jobs kubecost cost-analyzer/ -n kubecost -f "$file"
+            echo "### Waiting for cost-analyzer Pod readiness. ###"
+            kubectl wait -n kubecost --for=condition=ready pod --selector app.kubernetes.io/name=cost-analyzer --timeout=120s
+            echo "### Sleeping for 60 seconds longer. ###"
+            sleep 60
+            echo "### [DEBUG] Listing Pods in Kubecost Namespace. ###"
+            kubectl -n kubecost get pods
+            echo "### Performing helm tests ###"
+            helm test -n kubecost $(helm ls -n kubecost --all --short)
```
Comment (thomasvn) [inline .github/workflows/chart.yaml:130]: Could you help clarify what exactly `helm test` does once Kubecost is installed?

File: .github/workflows/chart.yaml [hunk 3]
```diff
@@ -14,22 +14,34 @@ jobs:
...
+      # Run `helm template` on the main values file plus all those in the ci directory.
       - name: Helm template
-        working-directory: ./cost-analyzer
-        run: helm template . --set global.prometheus.enabled=false --set global.grafana.enabled=false > full.yaml
+        run: |
+          helm template cost-analyzer/ -f cost-analyzer/values.yaml > full.yaml
+          directory="cost-analyzer/ci"
+          for file in "$directory"/*; do
+          if [ -f "$file" ]; then
+              helm template cost-analyzer/ -f "$file" >> full.yaml
```
Comment (thomasvn) [inline .github/workflows/chart.yaml:31]: If we are looping over the entire `cost-analzyer/ci` directory. Then rendering the templates and appending output to `full.yaml`

Is the expected behavior for `Kubeconform` to later fail if at least one of these `values.yaml` files have rendered incorrectly?

**Comment by thomasvn** [comment]
@chipzoller Really awesome! Glad that we can now test deployments of these common configurations upon each PR. Only a few minor comments here, mostly questions for clarity.

Also, is there any reason the `aggregator-values.yaml` file is so large? Could we remove all the excess and just include what's pertinent to aggregator?

File: .github/workflows/chart.yaml (hunk 1 above)
Comment (chipzoller) [reply .github/workflows/chart.yaml:113]: The tests here, as opposed to the previous one, run across the values in the `ci` directory. So the total tests include the one with the default `values.yaml` file plus the values files in the `ci` directory.

File: .github/workflows/chart.yaml (hunk 2 above)
Comment (chipzoller) [reply .github/workflows/chart.yaml:130]: Sure, although [here](https://helm.sh/docs/topics/chart_tests/) are the docs. At this point in time, `helm test` will only execute the one test case I've written. After a chart is deployed, running `helm test` will run the test which is included as part of the chart. It is invoked only when the `test` command is used.

File: .github/workflows/chart.yaml (hunk 3 above)
Comment (chipzoller) [reply .github/workflows/chart.yaml:31]: No, Kubeconform is only used to assess conformance of the final, rendered output against Kubernetes resource specs. For example, let's say you had `spec.foo: bar` in a Pod. While this may template correctly, it's not conformant to the Pod specification. A `helm template` would not catch this because the templating may occur successfully, but the final resource does not conform to the known Pod spec. Kubeconform is used to ensure that the (successfully) templated output are actually valid K8s resources.

**Comment by chipzoller** [comment]
Great questions, as always.

> Also, is there any reason the `aggregator-values.yaml` file is so large? Could we remove all the excess and just include what's pertinent to aggregator?

No particular reason, I suppose. I used an existing config in the PoC repo plus your suggestions (plus a couple of my own, ex., enabling network costs) to come up with this manifest. If we can slim it down that'd be ideal. Do you have any recommendations?

**Comment by thomasvn** [comment]
How about using these values for the Aggregator? ([`aggregator-minimal.yaml` internal repo](https://github.com/kubecost/clusters/blob/main/how-to/aggregator-minimum.yaml))

File: .github/workflows/chart.yaml (hunk 2 above)
Comment (thomasvn) [reply .github/workflows/chart.yaml:130]: Oh I see now, thanks for linking to docs!

**Comment by chipzoller** [comment]
Can I opt out of using an existing ServiceAccount? This will reduce the dependencies by one. It doesn't impact the ability to determine if aggregator works properly, correct?

**Comment by thomasvn** [comment]
Is this the ServiceAccount you're referring to? Aggregator shouldn't need it!

```
# when using managed identity/irsa, set the service account accordingly:
serviceAccount:
  create: false
  name: kubecost-irsa-sa
```

**Comment by chipzoller** [comment]
Can `kubecostProductConfigs.productKey.key` just remain as `YOUR_KEY` or does it have to be something valid?

**Comment by chipzoller** [comment]
Hey @thomasvn, can you help me refine this so we can get it in?

**Comment by thomasvn** [comment]
@chipzoller Thoughts on omitting `kubecostProductConfigs.productKey.key` from any CI tests?

Even if we put a dummy string here, it wouldn't be testing much functionality. And we also wouldn't want to be putting a valid key in these CI tests unless it's possible to store this secret outside git.

**Comment by chipzoller** [comment]
Omitting that sounds good, I just don't know what happens if the value isn't provided. Can you summarize the changes you'd like to see in this PR?

**Comment by jessegoodier** [comment]
by default, no value is needed. The majority of installs are free.

**Comment by chipzoller** [comment]
But can it be null or does it have to be `""`?

**Comment by thomasvn** [comment]
@chipzoller Happy to get this PR in ASAP! No major concerns with it. I think as a minor improvement we can use a much more minimal `aggregator-values.yaml`.

When I configured a minimal aggregator, these were the only configs I needed:

```yaml
# kubectl create secret generic federated-store --from-file=federated-store.yaml
# kubectl create secret generic cloud-integration --from-file=cloud-integration.json
kubecostAggregator:
  enabled: true
  cloudCost:
    enabled: true
  aggregatorStorage:
    storageRequest: 5Gi
  aggregatorDbStorage:
    storageRequest: 10Gi
kubecostModel:
  federatedStorageConfigSecret: federated-store
kubecostProductConfigs:
  cloudIntegrationSecret: cloud-integration
```

**Comment by chipzoller** [comment]
@thomasvn, made the changes. Please review.

**Review by thomasvn — APPROVED** [review]
